### ISSUE TITLE
Correct way to get result from argmax?

### ISSUE THREAD

**Issue by dougdew64** [ml-explore/mlx#214]
I'm attempting to teach myself the MLX api by re-implementing llama2.cpp atop MLX. In this screen shot of Xcode, my code is on the left, and the llama2.cpp code is on the right.

Unfortunately, it's not clear to me how to perform some MLX operations. For example, getting the result from calling the MLX argmax function. As shown in the screen shot, I'm receiving a compilation error for this line of code in which I attempt to de-reference an array iterator:

`next = *argmax(state.logits).begin();`

I'm looking through the repo for usage examples, but am so far finding none.

![image](https://github.com/ml-explore/mlx/assets/98898390/721ae608-79eb-4452-9db3-f154949d5d97)

**Comment by dougdew64** [comment]
I've also attempted this line:

next = argmax(state.logits).item<int>();

At least that line compiles, but then I receive this error:

validateComputeFunctionArguments:1056: failed assertion `Compute Function(argmax_float32): missing buffer binding at index 2 for shape[0].'

Debugging...

**Comment by dougdew64** [comment]
![image](https://github.com/ml-explore/mlx/assets/98898390/bb060f7d-d7db-4048-b325-acc914f33215)

**Comment by dougdew64** [comment]
That error message seems not surprising given that ArgReduce::eval_gpu erases the shape data before using the shape to set bytes at index 2 for the compute encoder. 

![image](https://github.com/ml-explore/mlx/assets/98898390/aa36dace-2479-4e17-882d-2c81c936c835)

**Comment by dougdew64** [comment]
I found this test code. Now I just need to figure out how to run the tests.

![image](https://github.com/ml-explore/mlx/assets/98898390/b67e62bd-fa5c-43a8-9705-2b1d1307109c)

**Comment by dougdew64** [comment]
Test failure(s):

![image](https://github.com/ml-explore/mlx/assets/98898390/ef92f62e-aa0d-4e32-a416-a4bc3523094f)

**Comment by awni** [comment]
Yea the CPP tests are a good way to see some usage. The line:

```
auto amax = argmax(a).item<int>()
```

Should work. The fact that it didn't for you suggests something is wrong further up. But let me know if you think there is a bug somewhere!

**Comment by dougdew64** [comment]
Thanks @awni. I'll continue to debug my code.

**Comment by dougdew64** [comment]
I'm able to reproduce the error when running the argmin//argmax tests which are in arg_reduce_tests.cpp: https://github.com/ml-explore/mlx/issues/220
